crates/maybenot-ffi/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+/// - Create it: [maybenot_start].
+/// - Feed it actions: [maybenot_on_event].
+/// - Stop it: [maybenot_stop].
+pub struct MaybenotFramework {
+    framework: Framework<Vec<Machine>>,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct MaybenotEvent {
+    pub event_type: MaybenotEventType,
+
+    /// The number of bytes that was sent or received.
+    pub xmit_bytes: u16,
+
+    /// The ID of the machine that triggered the event, if any.
+    pub machine: usize,
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct MaybenotDuration {
+    /// Number of whole seconds
+    pub secs: u64,
+
+    /// A nanosecond fraction of a second.
+    pub nanos: u32,
+}
+
+#[repr(u32)]
+#[derive(Debug, Clone, Copy)]
+#[allow(dead_code)]
+pub enum MaybenotEventType {
+    /// We sent a normal packet.
+    NonpaddingSent = 0,
+
+    /// We received a normal packet.
+    NonpaddingReceived = 1,
+
+    /// We send a padding packet.
+    PaddingSent = 2,
+
+    /// We received a padding packet.
+    PaddingReceived = 3,
+}
+
+#[repr(C, u32)]
+#[derive(Debug, Clone, Copy)]
+pub enum MaybenotAction {
+    Cancel {
+        /// The machine that generated the action.
+        machine: usize,
+    } = 0,
+
+    /// Send a padding packet.
+    InjectPadding {
+        /// The machine that generated the action.
+        machine: usize,
+
+        /// The time to wait before injecting a padding packet.
+        timeout: MaybenotDuration,
+
+        replace: bool,
+        bypass: bool,
+
+        /// The size of the padding packet.
+        size: u16,
+    } = 1,
+
+    BlockOutgoing {
+        /// The machine that generated the action.
+        machine: usize,
+
+        /// The time to wait before blocking.
+        timeout: MaybenotDuration,
+
+        replace: bool,
+        bypass: bool,
+
+        /// How long to block.
+        duration: MaybenotDuration,
+    } = 2,
+}
+
+impl MaybenotFramework {
+    fn start(
+        machines_str: &str,
+        max_padding_bytes: f64,
+        max_blocking_bytes: f64,
+        mtu: u16,
+    ) -> Result<Self, MaybenotResult> {
+        let machines: Vec<_> = machines_str
+            .lines()
+            .map(|line| line.trim())
+            .filter(|line| !line.is_empty())
+            .filter(|line| !line.starts_with('#'))
+            .map(Machine::from_str)
+            .collect::<Result<_, _>>()
+            .map_err(|_e| MaybenotResult::InvalidMachineString)?;
+
+        let framework = Framework::new(
+            machines,
+            max_padding_bytes,
+            max_blocking_bytes,
+            mtu,
+            Instant::now(),
+        )
+        .map_err(|_e| MaybenotResult::StartFramework)?;
+
+        Ok(MaybenotFramework { framework })
+    }
+
+    fn on_event(
+        &mut self,
+        event: MaybenotEvent,
+        actions: &mut [MaybeUninit<MaybenotAction>],
+    ) -> usize {
+        let num_actions = self
+            .framework
+            .trigger_events(&[convert_event(event)], Instant::now())
```

Comment:
Here the API is changed from "list of events" (Rust) to "single event" (C). We can leave it like this for now. But I suspect we might revisit this in the future to make the API as similar to Rust as possible, and also we might hit performance bottlenecks making us want to batch event processing (like the Rust code allows)

Reply:
Done

Reply:
Unfortunately there's now a bit more code to review... :pensive: